tapchannel/aux_invoice_manager_test.go

```diff
@@ -0,0 +1,571 @@
+					),
+				},
+			},
+			responses: []lndclient.InvoiceHtlcModifyResponse{
+				{
+					AmtPaid: 3_000_000,
+				},
+			},
+			buyQuotes: rfq.BuyAcceptMap{
+				dummyRfqID(31).Scid(): {
+					Peer:     testNodeID,
+					AskPrice: 1_000_000,
+				},
+			},
+		},
+		{
+			name: "asset invoice, not enough amt",
+			requests: []lndclient.InvoiceHtlcModifyRequest{
+				{
+					Invoice: &lnrpc.Invoice{
+						RouteHints: testRouteHints(),
+						ValueMsat:  3_000_000,
+					},
+					WireCustomRecords: newWireCustomRecords(
+						t, []*rfqmsg.AssetBalance{
+							rfqmsg.NewAssetBalance(
+								assetID(1), 3,
+							),
+						}, fn.Some(dummyRfqID(31)),
+					),
+					ExitHtlcAmt: 1234,
+				},
+			},
+			responses: []lndclient.InvoiceHtlcModifyResponse{
+				{
+					AmtPaid: 1_500_000,
+				},
+			},
+			buyQuotes: rfq.BuyAcceptMap{
+				dummyRfqID(31).Scid(): {
+					Peer:     testNodeID,
+					AskPrice: 500_000,
+				},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+
+		t.Logf("Running AuxInvoiceManager test case: %v", testCase.name)
+
+		// Instantiate mock rfq manager.
+		mockRfq := &mockRfqManager{
+			peerBuyQuotes:   testCase.buyQuotes,
+			localSellQuotes: testCase.sellQuotes,
+		}
+
+		done := make(chan bool)
+
+		// Instantiate mock htlc modifier.
+		mockModifier := &mockHtlcModifier{
+			requestQue:     testCase.requests,
+			expectedResQue: testCase.responses,
+			done:           done,
+			t:              t,
+		}
+
+		// Create the manager.
+		manager := NewAuxInvoiceManager(
+			&InvoiceManagerConfig{
+				ChainParams:         testChainParams,
+				InvoiceHtlcModifier: mockModifier,
+				RfqManager:          mockRfq,
+			},
+		)
+
+		err := manager.Start()
+		require.NoError(t, err)
+
+		// If the manager is not done processing the htlc modification
+		// requests within the specified timeout, assume this is a
+		// failure.
+		select {
+		case <-done:
+		case <-time.After(testTimeout):
+			t.Fail()
+		}
+	}
+}
+
+// genRandomRfqID generates a random rfqmsg.ID value.
+func genRandomRfqID(t *rapid.T) rfqmsg.ID {
+	var rfqID rfqmsg.ID
+	for i := 0; i < len(rfqID); i++ {
+		rfqID[i] = rapid.Byte().Draw(t, "scid_byte")
+		// rapid.SliceOfN(rapid.Byte(), len(rfqID), len(rfqID))
+	}
+	return rfqID
+}
+
+// genInvoice generates an invoice that may have a random amount, and may have
+// routing hints.
+func genInvoice(t *rapid.T) *lnrpc.Invoice {
+	res := &lnrpc.Invoice{}
+
+	// Do a random draw with a 1/8 chance of introducing an empty invoice.
+	nullInv := rapid.Uint16().Draw(t, "invoice_exists")
+	if nullInv%8 == 0 {
+		return res
+	}
+
+	// Generate a random invoice value.
+	valueMsat := rapid.Int64Range(
+		0, 100_000_000,
+	).Draw(t, "invoice_value_msat")
+
+	res.ValueMsat = valueMsat
+
+	// Do a random draw with a 1/8 chance to introduce empty routing hints.
+	nullHints := rapid.Uint16().Draw(t, "invoice_hints_exist")
+	if nullHints%8 == 0 {
+		return res
+	}
+
+	res.RouteHints = testRouteHints()
```

Comment:
I think we can also have a generator for route hints of a random length and with random content. Right now I don't think the tests cover route hints that exist but are not matches to the expected route hint. Nor does it cover cases of more than 1 hint.

Reply:
> Right now I don't think the tests cover route hints that exist but are not matches to the expected route hint.

We actually check that, but the mismatch is always generated on the rfq quote generation, it's the `"nodeID_mismatch_probability"` value

> Nor does it cover cases of more than 1 hint.

Extended code to have variable route/hop hint lengths